### DISCUSSION TITLE
Cannot read property 'gssp' of undefined

### DISCUSSION THREAD

**Discussion by jgabriele** [vercel/next.js#23911] [Help] [answered]
### What version of Next.js are you using?

10.1.3

### What version of Node.js are you using?

14.15.4

### What browser are you using?

Chrome

### What operating system are you using?

macOS

### How are you deploying your application?

next start

### Describe the Bug

Starting from `v10.0.5` my tests all fail because of the usage of `createRouter` in them. It used to work fine before that but now they throw the following error: `Cannot read property 'gssp' of undefined`.

I could reproduce this issue [with this sandbox](https://codesandbox.io/s/sweet-lumiere-z0f7h).

### Expected Behavior

Tests are not crashing when using `createRouter` in them, as per `v10.0.4`.

### To Reproduce

https://codesandbox.io/s/sweet-lumiere-z0f7h

**Comment by timneutkens** [comment]
`createRouter` and `RouterContext` are not public apis.

**Reply by jgabriele** [reply]
Hello @timneutkens, thanks for your reply :). How can I mock Next router with public apis in my tests?

**Reply by jantimon** [reply]
@timneutkens whenever we want to generate a snapshot for a Component with `<Link />` or `useRouter` jest crashes 💣.  
Could you please help us out and add a public api which creates a mocked RouterContext?

An api might look like this:

```tsx
import { RouterContextMock} from "next/mock";

<RouterContextMock asPath="/blog/[post]" query={{post: 12}}>
  // ..
</RouterContextMock>
```

This is a big issue for a lot of next users: https://github.com/vercel/next.js/discussions/23034

**Comment by leosuncin** [comment, marked as answer]
I made it works, I took me sometime to figure it out, but you need to:

1. Set `setupFilesAfterEnv` in jest config

```javascript
setupFilesAfterEnv: ['<rootDir>/setupTests.ts'],
```

https://github.com/leosuncin/expense-tracker/blob/cf8bf4d352f1e099b938b175c558aef8c2379121/jest.config.ts#L8

2. Add this line to that file

```javascript
self.__NEXT_DATA__ = {};
```

https://github.com/leosuncin/expense-tracker/blob/cf8bf4d352f1e099b938b175c558aef8c2379121/setupTests.ts#L7-L8

3. Use `createRouter` to create the mocked router

```
import { createRouter } from 'next/router';

const router = createRouter('/login', {}, '/login', {
      subscription: jest.fn().mockImplementation(Promise.resolve),
      initialProps: {},
      pageLoader: jest.fn(),
      Component: jest.fn(),
      App: jest.fn(),
      wrapApp: jest.fn(),
      isFallback: false,
});
```

I also tried to mock the router without using `createRouter` but that didn't work

4. Pass the mocked router to `RouterContext` along with all your providers

```
import { render } from '@testing-library/react';
import { ConnectedRouter } from 'connected-next-router';
import { RouterContext } from 'next/dist/next-server/lib/router-context';
import { Provider } from 'react-redux';

render(
    <RouterContext.Provider value={router}>
      <Provider store={store}>
        <ConnectedRouter>
          <Component2Test />
        </ConnectedRouter>
      </Provider>
    </RouterContext.Provider>,
  );
```

To avoid repetition I created a [custom render](https://testing-library.com/docs/react-testing-library/setup#custom-render) (personal note: find why when I use `wrapper` options in render fail with `React.createElement is undefined` only happens with Next.js and Typescript)

https://github.com/leosuncin/expense-tracker/blob/master/src/utils/testUtils.tsx

5. Write the tests
```typescript
import { screen, waitFor } from '@testing-library/react';
import user from '@testing-library/user-event';
import { enableFetchMocks } from 'jest-fetch-mock';
import { createRouter } from 'next/router';

import { makeStore } from '@app/app/store';
import { loginFactory, userFactory } from '@app/features/auth/authFactories';
import LoginForm from '@app/features/auth/LoginForm';
import { render } from '@app/utils/testUtils';

test('submit the form', async () => {
  const store = makeStore();
  const router = createRouter('/login', {}, '/login', {
    subscription: jest.fn().mockImplementation(Promise.resolve),
    initialProps: {},
    pageLoader: jest.fn(),
    Component: jest.fn(),
    App: jest.fn(),
    wrapApp: jest.fn(),
    isFallback: false,
  });
  const data = loginFactory.build();
  const result = userFactory.build({
    email: data.email,
    isAdmin: false,
  });

  const spyReplace = jest.spyOn(router, 'replace');
  fetchMock.mockResponseOnce(JSON.stringify(result), { status: 201 });
  render(<LoginForm />, store, router);

  user.type(screen.getByLabelText(/email/i), data.email);
  user.type(screen.getByLabelText(/password/i), data.password);
  user.click(screen.getByRole('button', { name: /login/i }));

  await waitFor(() => {
    expect(screen.getByRole('button', { name: /login/i })).toBeEnabled();
  });

  expect(store.getState().auth).toMatchObject(
    expect.objectContaining({
      isAuthenticated: true,
      isLoading: false,
      user: result,
    }),
  );
  expect(spyReplace).toHaveBeenCalledWith('/', undefined, undefined);
});
```

https://github.com/leosuncin/expense-tracker/blob/master/src/features/auth/LoginForm.spec.tsx

**Reply by jgabriele** [reply]
Wow thanks a lot 🙏! This `self.__NEXT_DATA__` in the setup file did the trick. I am having other issues now but I need to have a closer look.

Anyway it feels unsafe to me to have no choice but using private apis to mock the router part, @timneutkens if you know a better way of doing it, that would be awesome.
